time/src/time.rs

```diff
@@ -29,21 +29,152 @@ pub(crate) enum Padding {
 /// (either positive or negative).
 ///
 /// When comparing two `Time`s, they are assumed to be in the same calendar date.
-#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+#[derive(Clone, Copy)]
+// The alignment is forced to 8 as an heuristic for performances. This is not
+// required for safety.
+#[repr(C, align(8))]
 pub struct Time {
+    // The order of this struct's fields matter!
+    // Do not change them.
+
+    // Little endian version
+    #[cfg(target_endian = "little")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    nanosecond: u32,
+    #[cfg(target_endian = "little")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    second: u8,
+    #[cfg(target_endian = "little")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    minute: u8,
+    #[cfg(target_endian = "little")]
     #[allow(clippy::missing_docs_in_private_items)]
     hour: u8,
+    #[cfg(target_endian = "little")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    padding: Padding,
+
+    // Big endian version
+    #[cfg(target_endian = "big")]
+    #[allow(clippy::missing_docs_in_private_items)]
+    hour: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     minute: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     second: u8,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     nanosecond: u32,
+    #[cfg(target_endian = "big")]
     #[allow(clippy::missing_docs_in_private_items)]
     padding: Padding,
 }
 
+impl core::hash::Hash for Time {
+    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+        self.as_u64().hash(state)
+    }
+}
+
+impl PartialEq for Time {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_u64().eq(&other.as_u64())
+    }
+
+    #[allow(clippy::partialeq_ne_impl)]
+    fn ne(&self, other: &Self) -> bool {
+        self.as_u64().ne(&other.as_u64())
+    }
```

Comment:
Why implement various methods that have default implementations?

Reply:
Because for some default methods, the compiler didn't manage to generate the optimal assembly that this hand implementation gives. Rather than manually checking all of them and relying on compiler optimization, I provided them.

Reply:
Do you remember which one (or multiple) didn't generate optimal assembly? It may be something that can be tackled in the compiler. Either way, it's less than ideal to implement method calls that are effectively redundant.

Reply:
Well you were right to ask me to double check. I don't know if I did it differently this time, or I just did it wrong last time, but in any case, it seems to be ok without any unsafe code nor default impl now.